File: 1/sol.py
import numpy as np
from functools import reduce

# More general solution for part 2
# Pre-condition: len(inputs) >= 3
def general_sol(inputs, n):
    shifted = np.roll(inputs, -n)
    result = 0
    return sum(map(lambda x : (x > 0), (shifted - inputs)[:-n]))
#   for k in (shifted - inputs)[:-n]:
#       if k > 0:
#           result += 1
#   return result

# read file and start the solution
inputs = np.array([])
with open("input.txt", "r") as f:
    for lines in f:
        num = int(lines)
        inputs = np.append(inputs, num)

print("Part 1:", general_sol(inputs, 1))
print("Part 2:", general_sol(inputs, 3))

File: 1/sol.hs
import System.IO
import Text.Printf

--- Part 1 & Part 2 solution
generalSol :: Int -> [Int] -> Int
generalSol n lst
  | length lst <= n = 0
  | otherwise       = indicator + generalSol n (tail lst)
  where
    diff = (lst !! n) - (lst !! 0)
    indicator = if diff > 0 then 1 else 0

-- Read input and pass it on
main :: IO ()
main = do
    handle <- openFile "input.txt" ReadMode
    contents <- hGetContents handle

    -- Parsing
    let contentLines = lines contents
        inputs       = map read contentLines

    printf "Part 1: %v \n" (generalSol 1 inputs)
    printf "Part 2: %v \n" (generalSol 3 inputs)
